src/device/rcp/rsp/rsp_core.c

```diff
@@ -89,6 +93,60 @@ static void dma_sp_read(struct rsp_core* sp)
         post_framebuffer_write(&sp->dp->fb, dramaddr - length, length);
         dramaddr+=skip;
     }
+
+    /* schedule end of dma event */
+    cp0_update_count(sp->mi->r4300);
+    add_interrupt_event(&sp->mi->r4300->cp0, RSP_DMA_INT, (count * length) / 8);
+}
+
+static void fifo_push(struct rsp_core* sp, uint32_t dir)
+{
+    if (sp->regs[SP_DMA_BUSY_REG])
+    {
+        sp->fifo[1].dir = dir;
+        sp->fifo[1].length = dir == SP_DMA_READ ? sp->regs[SP_WR_LEN_REG] : sp->regs[SP_RD_LEN_REG];
+        sp->fifo[1].memaddr = sp->regs[SP_MEM_ADDR_REG];
+        sp->fifo[1].dramaddr = sp->regs[SP_DRAM_ADDR_REG];
+        sp->regs[SP_DMA_FULL_REG] = 1;
+        sp->regs[SP_STATUS_REG] |= SP_STATUS_DMA_FULL;
+    }
+    else
+    {
+        sp->fifo[0].dir = dir;
+        sp->fifo[0].length = dir == SP_DMA_READ ? sp->regs[SP_WR_LEN_REG] : sp->regs[SP_RD_LEN_REG];
+        sp->fifo[0].memaddr = sp->regs[SP_MEM_ADDR_REG];
+        sp->fifo[0].dramaddr = sp->regs[SP_DRAM_ADDR_REG];
+        sp->regs[SP_DMA_BUSY_REG] = 1;
+        sp->regs[SP_STATUS_REG] |= SP_STATUS_DMA_BUSY;
+
+        if (sp->fifo[0].dir == SP_DMA_READ)
+            dma_sp_read(sp, &sp->fifo[0]);
+        else
+            dma_sp_write(sp, &sp->fifo[0]);
+    }
+}
+
+static void fifo_pop(struct rsp_core* sp)
+{
+    if (sp->regs[SP_DMA_FULL_REG])
+    {
+        sp->fifo[0].dir = sp->fifo[1].dir;
+        sp->fifo[0].length = sp->fifo[1].length;
+        sp->fifo[0].memaddr = sp->fifo[1].memaddr;
+        sp->fifo[0].dramaddr = sp->fifo[1].dramaddr;
+        sp->regs[SP_DMA_FULL_REG] = 0;
+        sp->regs[SP_STATUS_REG] &= ~SP_STATUS_DMA_FULL;
+
+        if (sp->fifo[0].dir == SP_DMA_READ)
+            dma_sp_read(sp, &sp->fifo[0]);
+        else
+            dma_sp_write(sp, &sp->fifo[0]);
+    }
+    else
+    {
+        sp->regs[SP_DMA_BUSY_REG] = 0;
```

Comment:
You may emit a warning if SP_DMA_BUSY_REG == 0

Reply:
I don't think this would be possible, DMA_BUSY_REG is set in fifo_push

Reply:
Indeed, given how the fifo_{push,pop} is currently called, but it was more for "symmetry" : we checked for full fifo in fifo_push, so I suggested to also check for empty fifo in fifo_pop.

But either way is fine by me.